nomos-da/network/core/src/protocols/replication/behaviour_v2.rs

```diff
@@ -0,0 +1,677 @@
+/// member of the addressed `SubnetworkId`.
+pub struct ReplicationBehaviour<Membership> {
+    /// Local peer Id, related to the libp2p public key
+    local_peer_id: PeerId,
+    /// Underlying stream behaviour
+    stream_behaviour: libp2p_stream::Behaviour,
+    /// Used to open new outbound streams from the stream behaviour
+    control: Control,
+    /// Provides inbound streams that are accepted by the stream behaviour
+    incoming_streams: IncomingStreams,
+    /// Holds tasks for reading messages from incoming streams
+    incoming_tasks: FuturesUnordered<IncomingTask>,
+    /// Membership handler, membership handles the subsets logics on who is
+    /// where in the nomos DA subnetworks
+    membership: Membership,
+    /// Currently connected peers
+    connected: HashSet<PeerId>,
+    /// Pending outbound messages are stored here, they are then consumed by
+    /// each respective outbound stream
+    pending_outbound: PendingOutbound,
+    /// Indicates which outbound streams are currently idle or busy
+    outbound_streams: HashMap<PeerId, StreamState>,
+    /// Holds tasks for writing messages to the outbound streams
+    outbound_tasks: FuturesUnordered<OutboundTask>,
+    /// Seen messages cache holds a record of seen messages
+    seen_message_cache: IndexSet<(Vec<u8>, SubnetworkId)>,
+    /// Waker that handles polling
+    waker: Option<Waker>,
+}
+
+impl<Membership> ReplicationBehaviour<Membership> {
+    pub fn new(peer_id: PeerId, membership: Membership) -> Self {
+        let stream_behaviour = libp2p_stream::Behaviour::new();
+        let mut control = stream_behaviour.new_control();
+        let incoming_streams = control
+            .accept(REPLICATION_PROTOCOL)
+            .expect("A unique protocol can be accepted only once");
+        let incoming_tasks = FuturesUnordered::new();
+        let outbound_tasks = FuturesUnordered::new();
+
+        Self {
+            local_peer_id: peer_id,
+            stream_behaviour,
+            control,
+            incoming_streams,
+            incoming_tasks,
+            membership,
+            connected: HashSet::default(),
+            pending_outbound: PendingOutbound::default(),
+            outbound_streams: HashMap::default(),
+            outbound_tasks,
+            seen_message_cache: IndexSet::default(),
+            waker: None,
+        }
+    }
+
+    pub fn update_membership(&mut self, membership: Membership) {
+        self.membership = membership;
+    }
+}
+
+impl<M> ReplicationBehaviour<M>
+where
+    M: MembershipHandler<NetworkId = SubnetworkId, Id = PeerId>,
+{
+    /// Check if some peer membership lies in at least a single subnetwork that
+    /// the local peer is a member too.
+    fn is_neighbour(&self, peer_id: &PeerId) -> bool {
+        self.membership
+            .membership(&self.local_peer_id)
+            .intersection(&self.membership.membership(peer_id))
+            .count()
+            > 0
+    }
```

Comment:
I thought something like this was already in membership. If it is not, we should probably make it part of the membership API. This is because now membership is a local copy of the data, but at some point it is probably gonna be a service itself. We can left it like that now, and open an issue to track this and do it later on.